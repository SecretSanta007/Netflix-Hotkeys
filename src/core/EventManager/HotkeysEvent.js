const HotkeysEventType = Object.freeze({
  PLAY_SPEED_FASTEST: 'PLAY_SPEED_FASTEST',
  PLAY_SPEED_NORMAL: 'PLAY_SPEED_NORMAL',
  PLAY_SPEED_SLOWEST: 'PLAY_SPEED_SLOWEST',
  SUBTITLE_ON: 'SUBTITLE_ON',
  SUBTITLE_OFF: 'SUBTITLE_OFF',
  AUDIO_CHANGE_ENGLISH: 'AUDIO_CHANGE_ENGLISH',
  AUDIO_CHANGE_ORIGINAL: 'AUDIO_CHANGE_ORIGINAL',
})

class HotkeysEvent {
  /**
   * Type of the event.
   * @type {HotkeysEventType|null}
   */
  type = null

  /**
   * Creates a new Hotkeys Event.
   * @param {HotkeysEventType} type
   */
  constructor(type) {
    this.type = type
  }
}

export { HotkeysEvent, HotkeysEventType }
