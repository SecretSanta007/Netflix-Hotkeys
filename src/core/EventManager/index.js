export { default as eventManager } from './EventManager'
export { HotkeysEvent, HotkeysEventType } from './HotkeysEvent'
