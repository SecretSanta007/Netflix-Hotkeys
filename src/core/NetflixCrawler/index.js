const Page = Object.freeze({
  USER_SELECTION: 'USER_SELECTION',
  PASSWORD_INPUT: 'PASSWORD_INPUT',
  PASSWORD_INPUT_WRONG: 'PASSWORD_INPUT_WRONG',
  VIDEO_PLAYER: 'VIDEO_PLAYER',
  UNKNOWN: 'UNKNOWN',
})

const mainObserver = () => {
  console.info('[🕸️] Netflix Crawler: Main Observer triggered')
  _callListeners()
}

const netflixObserver = new MutationObserver(mainObserver)

function start() {
  netflixObserver.observe(document.documentElement || document.body, {
    childList: true,
    subtree: true,
  })
}

function stop() {
  netflixObserver.disconnect()
}

/** @type {function[]} */
const _changeListeners = []

/**
 * Calls the callback when the netflix changes DOM.
 * @param {function} callback
 */
function addChangeListener(callback) {
  _changeListeners.push(callback)
}

function _callListeners() {
  for (const listener of _changeListeners) {
    listener()
  }
}

const guessPage = () => {
  const firstHeader = document.getElementsByTagName('h1')[0]?.innerText
  if (firstHeader === "Who's watching?") {
    return Page.USER_SELECTION
  }

  if (firstHeader === 'Enter your PIN to access this profile.') {
    return Page.PASSWORD_INPUT
  }

  if (firstHeader === 'Whoops, wrong PIN. Please try again.') {
    return Page.PASSWORD_INPUT_WRONG
  }

  if (controls.videoPlayer) {
    return Page.VIDEO_PLAYER
  }

  return Page.UNKNOWN
}

// Lazily loaded controls
let controls = {
  get page() {
    return guessPage()
  },

  /** @type {NodeListOf<HTMLElement>|null} */
  get profileNameEls() {
    return document.querySelectorAll('span.profile-name')
  },

  /** @type {HTMLElement|null} */
  get profilePasswordInput() {
    return document.querySelector("input[data-uia='pin-number-0']")
  },

  /** @type {HTMLElement|null} */
  get videoPlayer() {
    return document.querySelector("div[data-uia='watch-video']")
  },

  /** @type {HTMLElement|null} */
  get backSeekButton() {
    return document.querySelector("button[data-uia='control-back10']")
  },

  /** @type {HTMLElement|null} */
  get forwardSeekButton() {
    return document.querySelector("button[data-uia='control-forward10']")
  },

  /** @type {HTMLElement|null} */
  get skipIntroButton() {
    return document.querySelector("button[data-uia='player-skip-intro']")
  },

  /** @type {HTMLElement|null} */
  get skipRecapButton() {
    return document.querySelector("button[data-uia='player-skip-recap']")
  },

  /** @type {HTMLElement|null} */
  get skipToNextEpisodeButton() {
    return document.querySelector("button[data-uia='next-episode-seamless-button']")
  },

  /** @type {HTMLElement|null} */
  get nextEpisodeButton() {
    return document.querySelector("button[data-uia='control-next']")
  },

  /** @type {HTMLElement|null} */
  get timeline() {
    return document.querySelector("div[data-uia='timeline']")
  },

  /** @type {HTMLElement|null} */
  get controlAudioSubtitleButton() {
    return document.querySelector("button[data-uia='control-audio-subtitle']")
  },

  /** @type {HTMLElement|null} */
  get controlAudioSubtitleMenu() {
    return document.querySelector("div[data-uia='selector-audio-subtitle']")
  },

  /** @type {HTMLElement[]} */
  get englishAudioOption() {
    return document.querySelector("li[data-uia*='audio-item'][data-uia*='english' i]")
  },

  /** @type {HTMLElement|null} */
  get originalAudioOption() {
    return document.querySelector("li[data-uia*='audio-item'][data-uia*='original' i]")
  },

  /** @type {HTMLElement|null} */
  get selectedAudioOption() {
    return document.querySelector("li[data-uia*='audio-item-selected']")
  },

  /** @type {HTMLElement|null} */
  get englishSubtitleOption() {
    return document.querySelector("li[data-uia*='subtitle-item'][data-uia*='english' i]")
  },

  /** @type {HTMLElement|null} */
  get offSubtitleOption() {
    return document.querySelector("li[data-uia*='subtitle-item'][data-uia*='Off' i]")
  },

  /** @type {HTMLElement|null} */
  get selectedSubtitleOption() {
    return document.querySelector("li[data-uia*='subtitle-item-selected']")
  },

  /** @type {HTMLElement|null} */
  get speedControlButton() {
    return document.querySelector("button[data-uia='control-speed']")
  },

  /** @type {HTMLElement|null} */
  get speedControlMenu() {
    return document.querySelector("div[data-uia='playback-speed']")
  },

  /** @type {HTMLElement[]} */
  get speedControlOptions() {
    return Array.from(document.querySelectorAll("div[data-uia*='playback-speed-item']"))
  },

  /** @type {HTMLElement|null} */
  get slowestSpeedOption() {
    return this.speedControlOptions[0]
  },

  /** @type {HTMLElement|null} */
  get normalSpeedOption() {
    return this.speedControlOptions.at(Math.floor(this.speedControlOptions.length / 2))
  },

  /** @type {HTMLElement|null} */
  get fastestSpeedOption() {
    return this.speedControlOptions.at(this.speedControlOptions.length - 1)
  },
}

export default {
  Page,
  start,
  stop,
  addChangeListener,
  controls,
}

export { Page }
