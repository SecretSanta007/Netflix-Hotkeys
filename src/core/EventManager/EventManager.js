import { HotkeysEvent } from './HotkeysEvent'

class EventManager {
  /**
   * Publish the passed event.
   * @param {HotkeysEvent} event
   */
  publish(event) {
    this.#dispatchEvent(event)
  }

  /** @type {Function[]} */
  #listeners = []

  /**
   * Add an event listener.
   * @param {Function} callback
   */
  addEventListener(callback) {
    this.#listeners.push(callback)
  }

  /**
   * Remove an event listener.
   * @param {Function} callback
   */
  removeEventListener(callback) {
    this.#listeners = this.#listeners.filter((listener) => listener !== callback)
  }

  /**
   * Dispatch an event.
   * @param {HotkeysEvent} event
   */
  #dispatchEvent(event) {
    this.#listeners.forEach((listener) => listener(event))
  }

  /**
   * Resets the Manager.
   *
   * NOTE: Used for testing purposes.
   */
  reset() {
    this.#listeners = []
  }
}

export default new EventManager()
