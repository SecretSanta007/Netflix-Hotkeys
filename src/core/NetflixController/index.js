import { preferences } from '../Preferences'
import NetflixCrawler, { Page } from '../NetflixCrawler'
import Executor from './Executor'

/**
 * User Selection Observer
 * Observes the user selection page and selects the default user if auto login is enabled.
 */
const userSelectionPageObserver = () => {
  let defaultUsernameEl = undefined
  const usernames = []
  for (const usernameEl of NetflixCrawler.controls.profileNameEls) {
    usernames.push(usernameEl.innerText)
    if (usernameEl.innerText === preferences.defaultUsername) {
      defaultUsernameEl = usernameEl
    }
  }

  preferences.usernameList = usernames
  defaultUsernameEl?.click()
}

/**
 * Password Input Observer
 * Observes the password input page and enters the default password if auto login is enabled.
 * Detail: Copy pastes the password to the first input field.
 */
const passwordInputPageObserver = () => {
  const password = preferences.profilePassword
  if (!password && password.length !== 4 && isNaN(password)) {
    return
  }

  const clipboardData = new DataTransfer()
  clipboardData.setData('text/plain', password)
  NetflixCrawler.controls.profilePasswordInput.dispatchEvent(
    new ClipboardEvent('paste', { clipboardData: clipboardData, bubbles: true }),
  )
}

/**
 * Video Player Observer
 * Observes the video player page and skips intro and auto plays next episode if auto skip is enabled.
 * Also, stores the reference to the buttons for seeking forward and backward.
 */
const videoPlayerPageObserver = () => {
  if (preferences.isPowerSkipEnabled) {
    NetflixCrawler.controls.skipIntroButton?.click()
    NetflixCrawler.controls.skipToNextEpisodeButton?.click()
  }
}

NetflixCrawler.addChangeListener(() => {
  const page = NetflixCrawler.controls.page
  if (page === Page.USER_SELECTION) {
    userSelectionPageObserver()
  } else if (page === Page.PASSWORD_INPUT) {
    passwordInputPageObserver()
  } else if (page === Page.VIDEO_PLAYER) {
    videoPlayerPageObserver()
  }
})

function start() {
  NetflixCrawler.start()
}

function stop() {
  NetflixCrawler.stop()
}

function seekForward() {
  if (!preferences.isSlowSeekEnabled) {
    return
  }

  Executor.executeOrAddToQueue(() => {
    return clickIfButtonExists(NetflixCrawler.controls.forwardSeekButton)
  })
}

function seekBackward() {
  if (!preferences.isSlowSeekEnabled) {
    return
  }

  Executor.executeOrAddToQueue(() => {
    return clickIfButtonExists(NetflixCrawler.controls.backSeekButton)
  })
}

function jumpToNextEpisode() {
  if (!preferences.isStartNextEpisodeEnabled) {
    return
  }
  Executor.executeOrAddToQueue(() => {
    return clickIfButtonExists(NetflixCrawler.controls.nextEpisodeButton)
  })
}

function startOverEpisode() {
  if (!preferences.isStartOverEpisodeEnabled) {
    return
  }

  Executor.executeOrAddToQueue(() => {
    if (!NetflixCrawler.controls.timeline) {
      return false
    }

    // After much debugging and looking through all things like "chrome.debugger" and "puppeteer" following logic was found to work.
    // For some weird reason we need to dispatch both mouse down and mouse up events to seek to the beginning of the episode.
    // NOTE: This has more probability of breaking in future.
    const timelineRect = NetflixCrawler.controls.timeline.getBoundingClientRect()
    const x = timelineRect.left
    const y = timelineRect.top + timelineRect.height * 0.5

    NetflixCrawler.controls.timeline.dispatchEvent(
      new MouseEvent('mousedown', {
        clientX: x,
        clientY: y,
        bubbles: true,
      }),
    )

    NetflixCrawler.controls.timeline.dispatchEvent(
      new MouseEvent('mouseup', {
        clientX: x,
        clientY: y,
        bubbles: true,
      }),
    )

    return true
  })
}

function callIfNetflixHotkeysEnabled(func) {
  return () => {
    if (preferences.isNetflixHotkeysEnabled) {
      func()
    }
  }
}

function clickIfButtonExists(button) {
  if (button) {
    button.click()
    return true
  }
  return false
}

export default {
  start,
  stop,
  seekForward: callIfNetflixHotkeysEnabled(seekForward),
  seekBackward: callIfNetflixHotkeysEnabled(seekBackward),
  jumpToNextEpisode: callIfNetflixHotkeysEnabled(jumpToNextEpisode),
  startOverEpisode: callIfNetflixHotkeysEnabled(startOverEpisode),
}
