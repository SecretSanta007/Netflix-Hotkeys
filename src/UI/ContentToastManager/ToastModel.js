import { HotkeysEvent, HotkeysEventType } from '../../core/EventManager'
import SpeedOutlined from '@mui/icons-material/SpeedOutlined'
import ClosedCaptionOffOutlinedIcon from '@mui/icons-material/ClosedCaptionOffOutlined'
import ClosedCaptionDisabledOutlinedIcon from '@mui/icons-material/ClosedCaptionDisabledOutlined'
import AudiotrackOutlinedIcon from '@mui/icons-material/AudiotrackOutlined'

const DEFAULT_AUTO_CLOSE_DELAY = 850

class ToastModel {
  icon = null
  text = null
  shouldAutoClose = true
  autoCloseDelay = DEFAULT_AUTO_CLOSE_DELAY

  constructor(icon, text, shouldAutoClose = true, autoCloseDelay = DEFAULT_AUTO_CLOSE_DELAY) {
    this.icon = icon
    this.text = text
    this.shouldAutoClose = shouldAutoClose
    this.autoCloseDelay = shouldAutoClose ? autoCloseDelay : null
  }

  /**
   * Create a ToastModel from a HotkeysEvent.
   * @param {HotkeysEvent} event
   * @returns {ToastModel|null}
   */
  static fromEvent(event) {
    if (!event || !event.type) {
      return null
    }

    if (event.type === HotkeysEventType.PLAY_SPEED_SLOWEST) {
      return new ToastModel(SpeedOutlined, '0.5x', false)
    } else if (event.type === HotkeysEventType.PLAY_SPEED_FASTEST) {
      return new ToastModel(SpeedOutlined, '1.5x', false)
    } else if (event.type === HotkeysEventType.SUBTITLE_ON) {
      return new ToastModel(ClosedCaptionOffOutlinedIcon, 'Subtitles on')
    } else if (event.type === HotkeysEventType.SUBTITLE_OFF) {
      return new ToastModel(ClosedCaptionDisabledOutlinedIcon, 'Subtitles off')
    } else if (event.type === HotkeysEventType.AUDIO_CHANGE_ENGLISH) {
      return new ToastModel(AudiotrackOutlinedIcon, 'Audio changed to English')
    } else if (event.type === HotkeysEventType.AUDIO_CHANGE_ORIGINAL) {
      return new ToastModel(AudiotrackOutlinedIcon, 'Audio changed to Original')
    }

    return null
  }
}

export default ToastModel
