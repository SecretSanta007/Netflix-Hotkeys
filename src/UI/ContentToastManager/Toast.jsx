import { Icon, Paper, Snackbar, Stack, Typography } from '@mui/material'
import React from 'react'
import ToastModel from './ToastModel'

/**
 * @param {Object} props
 * @param {ToastModel} props.model
 * @param {() => void} props.onClose
 */
function Toast({ model, onClose }) {
  if (!model) {
    return null
  }

  return (
    <Snackbar
      open={true}
      anchorOrigin={{ horizontal: 'center', vertical: 'top' }}
      onClose={onClose}
      autoHideDuration={model.autoCloseDelay}
    >
      <Paper
        elevation={0}
        sx={{ paddingX: 2, borderRadius: 2, bgcolor: 'rgba(1,1,1,0.5)', color: 'primary.contrastText' }}
      >
        <Stack direction="row" alignItems="center" spacing={1}>
          {model.icon && <Icon component={model.icon} sx={{ fontSize: '34px' }} />}
          {model.text && <Typography fontSize={32}>{model.text}</Typography>}
        </Stack>
      </Paper>
    </Snackbar>
  )
}

export default Toast
