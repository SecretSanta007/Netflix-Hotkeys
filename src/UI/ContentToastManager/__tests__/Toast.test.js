import Toast from '../Toast'
import ToastModel from '../ToastModel'
import { HotkeysEvent, HotkeysEventType } from '../../../core/EventManager'
import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

jest.useFakeTimers()

describe('Toast render for', () => {
  it('null model', () => {
    const { container } = render(<Toast model={null} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toBeNull()
  })

  it('Play fastest event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.PLAY_SPEED_FASTEST))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('1.5x')).toBeInTheDocument()
  })

  it('Play slowest event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.PLAY_SPEED_SLOWEST))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('0.5x')).toBeInTheDocument()
  })

  it('Play normal event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.PLAY_SPEED_NORMAL))
    const { container } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toBeNull()
  })

  it('Audio changed to english event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.AUDIO_CHANGE_ENGLISH))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('Audio changed to English')).toBeInTheDocument()
  })

  it('Audio changed to original event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.AUDIO_CHANGE_ORIGINAL))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('Audio changed to Original')).toBeInTheDocument()
  })

  it('Subtitle on event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.SUBTITLE_ON))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('Subtitles on')).toBeInTheDocument()
  })

  it('Subtitle off event model', () => {
    const model = ToastModel.fromEvent(new HotkeysEvent(HotkeysEventType.SUBTITLE_OFF))
    const { container, getByText } = render(<Toast model={model} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText('Subtitles off')).toBeInTheDocument()
  })
})

describe('Toast onClose', () => {
  it('should be called', () => {
    const model = new ToastModel(null, null, /** shouldAutoClose = */ true)
    const onClose = jest.fn()
    const {} = render(<Toast model={model} onClose={onClose} />)

    act(() => {
      jest.advanceTimersByTime(1000)
    })

    expect(onClose).toHaveBeenCalled()
  })

  it('should not be called', () => {
    const model = new ToastModel(null, null, /** shouldAutoClose = */ false)
    const onClose = jest.fn()
    const {} = render(<Toast model={model} onClose={onClose} />)

    act(() => {
      jest.advanceTimersByTime(1000)
    })

    expect(onClose).not.toHaveBeenCalled()
  })
})
