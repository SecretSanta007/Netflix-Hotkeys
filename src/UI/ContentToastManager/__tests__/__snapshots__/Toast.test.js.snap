// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast render for Audio changed to english event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="AudiotrackOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6zm-2 16c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        Audio changed to English
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for Audio changed to original event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="AudiotrackOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 3v10.55c-.59-.34-1.27-.55-2-.55-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4V7h4V3h-6zm-2 16c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        Audio changed to Original
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for Play fastest event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="SpeedOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m20.38 8.57-1.23 1.85a8 8 0 0 1-.22 7.58H5.07A8 8 0 0 1 15.58 6.85l1.85-1.23A10 10 0 0 0 3.35 19a2 2 0 0 0 1.72 1h13.85a2 2 0 0 0 1.74-1 10 10 0 0 0-.27-10.44z"
        />
        <path
          d="M10.59 15.41a2 2 0 0 0 2.83 0l5.66-8.49-8.49 5.66a2 2 0 0 0 0 2.83z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        1.5x
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for Play normal event model 1`] = `null`;

exports[`Toast render for Play slowest event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="SpeedOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m20.38 8.57-1.23 1.85a8 8 0 0 1-.22 7.58H5.07A8 8 0 0 1 15.58 6.85l1.85-1.23A10 10 0 0 0 3.35 19a2 2 0 0 0 1.72 1h13.85a2 2 0 0 0 1.74-1 10 10 0 0 0-.27-10.44z"
        />
        <path
          d="M10.59 15.41a2 2 0 0 0 2.83 0l5.66-8.49-8.49 5.66a2 2 0 0 0 0 2.83z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        0.5x
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for Subtitle off event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="ClosedCaptionDisabledOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 10c0-.55.45-1 1-1h3c.55 0 1 .45 1 1v1h-1.5v-.5h-2v1L13 10zm3.5 3.5 1.21 1.21c.18-.19.29-.44.29-.71v-1h-1.5v.5zM8.83 6H19v10.17l1.98 1.98c0-.05.02-.1.02-.16V6c0-1.1-.9-2-2-2H6.83l2 2zm10.95 16.61L17.17 20H5c-1.11 0-2-.9-2-2V6c0-.05.02-.1.02-.15L1.39 4.22 2.8 2.81l18.38 18.38-1.4 1.42zM7.5 13.5h2V13h.67l-2.5-2.5H7.5v3zm7.67 4.5L11 13.83V14c0 .55-.45 1-1 1H7c-.55 0-1-.45-1-1v-4c0-.32.16-.59.4-.78L5 7.83V18h10.17z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        Subtitles off
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for Subtitle on event model 1`] = `
<div
  class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-zzms1-MuiSnackbar-root"
  role="presentation"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-itv364-MuiPaper-root"
    direction="down"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="MuiStack-root css-1q36rmd-MuiStack-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1mwz4t-MuiSvgIcon-root-MuiIcon-root"
        data-testid="ClosedCaptionOffOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 4H5c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 14H5V6h14v12z"
        />
        <path
          d="M7 15h3c.55 0 1-.45 1-1v-1H9.5v.5h-2v-3h2v.5H11v-1c0-.55-.45-1-1-1H7c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1zm7 0h3c.55 0 1-.45 1-1v-1h-1.5v.5h-2v-3h2v.5H18v-1c0-.55-.45-1-1-1h-3c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1z"
        />
      </svg>
      <p
        class="MuiTypography-root MuiTypography-body1 css-fldieq-MuiTypography-root"
      >
        Subtitles on
      </p>
    </div>
  </div>
</div>
`;

exports[`Toast render for null model 1`] = `null`;
